addons/pid/fnc_update.sqf

```diff
@@ -0,0 +1,103 @@
+
+Description:
+    Updates a PID controller and returns the controller output.
+
+Parameters:
+    _pid    - a pid controller <LOCATION>
+    _value  - the measured value for this update <NUMBER>
+    _time   - the time for which the value was measured <NUMBER>
+              (Default: CBA_missionTime)
+
+Returns:
+    The output of the controller <NUMBER>
+
+Examples:
+    (begin example)
+        private _delta = [_pid, 6, CBA_missionTime] call CBA_pid_fnc_update;
+    (end)
+
+Author:
+    tcvm
+---------------------------------------------------------------------------- */
+SCRIPT(update);
+params [
+    ["_pid", locationNull, [locationNull]],
+    ["_value", 0, [0]],
+    ["_time", CBA_missionTime, [0]]
+];
+
+if (isNull _pid) exitWith { 0 };
+private _error = [_value, _pid getVariable [QGVAR(setpoint), 0]] call (_pid getVariable [QGVAR(errorFunction), FUNC(error_linear)]);
+private _history = _pid getVariable [QGVAR(history), []];
+private _maxHistoryLength = _pid getVariable [QGVAR(historyLength), DEFAULT_HISTORY_LENGTH];
+
+if (_history isNotEqualTo []) then {
+    // We need a continuous function, so if two measurements are for the same measured time we ignore the new measurements
+    // This also gets rid of any potential divide-by-zero errors due to the stride being 0
+    // The inverse must also be monotonic, so we assume time is always increasing
+    (_history select -1) params ["_x"];
+    if (_time - _x > 0) then {
+        _history pushBack [_time, _error];
+    };
+} else {
+    _history pushBack [_time, _error];
+};
+
+if (count _history > _maxHistoryLength) then {
+    _history deleteAt 0;
+};
+_pid setVariable [QGVAR(history), _history];
+
+private _derivative = 0;
+switch (true) do {
+    case (count _history == 2): {
+        private _x0 = _history select 0;
+        private _x1 = _history select 1;
+
+        private _stride = (_x1 select 0) - (_x0 select 0);
+        if (_stride == 0) then { break };
+
+        _derivative = ((_x1 select 1) - (_x0 select 1)) / _stride;
+    };
+    case (count _history >= 3): {
+        private _xn0 = _history select -1;
+        private _xn1 = _history select -2;
+        private _xn2 = _history select -3;
+        private _stride0 = (_xn0 select 0) - (_xn1 select 0);
+        private _stride1 = (_xn1 select 0) - (_xn2 select 0);
+        private _stride = _stride0 + _stride1;
+        if (_stride == 0) then { break };
+
+        private _sum = -3 * (_xn2 select 1) + 4 * (_xn1 select 1) - (_xn0 select 1);
+        _derivative = _sum / _stride;
+    };
+    default {};
+};
+
+private _integral = 0;
+private _error_prev = 0;
+private _tn_prev = 0;
+{
+    _x params ["_tn", "_error"];
+    if (_forEachIndex > 0) then {
+        // Trapezoidal rule for numerical integration
+        // (1 / 2) * (b - a) * (f(a) + f(b))
+        // O(h^2) error
+        private _stride = _tn - _tn_prev;
+        private _sum = _error + _error_prev;
+        _integral = _integral + 0.5 * _stride * _sum;
```

Comment:
There is no need for trapezoidal rule. Additionally this method of calculating the integral prevents from updating the PID parameters.

Reply:
How does this prevent the parameters from being updated? And although this method of estimating the integral is not needed, it isn’t meaningfully less efficient than a Reimann sum to change. It has a better approximation error, and isn’t meaningfully less efficient due to instructions